# [LAB 04] 데이터 다루기 - 5. 데이터 병합

## #01. 준비작업 

### 1. 패키지 참조

In [59]:
from jussam import load_data
from pandas import DataFrame 
from pandas import merge, concat

### 2. 첫번째 샘플 데이터 가져오기 

In [60]:
고객 = load_data("customer_no_name")
고객

📚 데이터프레임 병합 실습을 위한 고객번호와 고객 이름에 대한 가상 데이터(인덱스와 메타데이터 없음)


,고객번호,이름
0,1001,둘리
1,1002,도우너
2,1003,또치
3,1004,길동
4,1005,희동
5,1006,마이콜
6,1007,영희


### 3. 두번째 샘플 데이터 가져오기

In [61]:
매출 = load_data("customer_no_pay")
매출

📚 데이터프레임 병합 실습을 위한 고객번호와 매출액에 대한 가상 데이터(인덱스와 메타데이터 없음)


,고객번호,금액
0,1001,10000
1,1001,20000
2,1005,15000
3,1006,5000
4,1008,100000
5,1001,30000


## #02. 열 기준 병합

### 1. 일치하는 데이터끼리 병합

In [62]:
merge(고객, 매출)

,고객번호,이름,금액
0,1001,둘리,10000
1,1001,둘리,20000
2,1001,둘리,30000
3,1005,희동,15000
4,1006,마이콜,5000


### 2. 왼쪽 데이터 프레임을 기준으로 병합 

In [63]:
merge(고객, 매출, how = "left")

,고객번호,이름,금액
0,1001,둘리,10000.000
1,1001,둘리,20000.000
2,1001,둘리,30000.000
3,1002,도우너,NaN
4,1003,또치,NaN
5,1004,길동,NaN
6,1005,희동,15000.000
7,1006,마이콜,5000.000
8,1007,영희,NaN


### 3. 오른쪽 데이터 프레임을 기준으로 병합

In [64]:
merge(고객, 매출, how = "right")    

,고객번호,이름,금액
0,1001,둘리,10000
1,1001,둘리,20000
2,1005,희동,15000
3,1006,마이콜,5000
4,1008,NaN,100000
5,1001,둘리,30000


### 4. 모든 데이터의 교차 병합(합집합)

In [65]:
merge(고객, 매출, how="outer")

,고객번호,이름,금액
0,1001,둘리,10000.000
1,1001,둘리,20000.000
2,1001,둘리,30000.000
3,1002,도우너,NaN
4,1003,또치,NaN
5,1004,길동,NaN
6,1005,희동,15000.000
7,1006,마이콜,5000.000
8,1007,영희,NaN
9,1008,NaN,100000.000


### 03. 병합 기준 열 지정하기 

### 1. 샘플 데이터 가져오기 

In [66]:
cd1 = load_data("customer_data1")
cd1

📚 데이터 병합용 실습 데이터 (1) (인덱스/메타데이터 없음)


,고객명,날짜,데이터
0,민수,2018-01-01,20000
1,수영,2018-01-01,100000


In [67]:
cd2 = load_data("customer_data2")
cd2

📚 데이터 병합용 실습 데이터 (2) (인덱스/메타데이터 없음)


,고객명,데이터
0,민수,21세
1,수영,20세


### 2. 기본 병합 
- 두 데이터 프레임에서 이름이 같은 열은 모두 키가 된다.
- 샘플 데이터에서는 '데이터'라는 이름의 변수가 cd1은 int, cd2는 str 타입이므로 병합기준을 충족하지 않는다. 
- 그러므로 아래 코드는 에러

In [68]:
# You are trying to merge on int64 and str columns for key '데이터'. If you wish to proceed you should use pd.concat
# Output is truncated. View as a scrollable element or open in a text editor. Adjust cell output settings...
# merge(cd1, cd2)

### 3. 병합 기준 설정 
- on 파라미터로 병합의 기준이 되는 열을 지정해야 한다. 
- 병합 기준 열이 아니면서 이름이 같은 열에는 _x 또는 _y와 같은 접미사가 붙는다.

In [69]:
tmp = merge(cd1, cd2, on='고객명')
tmp

,고객명,날짜,데이터_x,데이터_y
0,민수,2018-01-01,20000,21세
1,수영,2018-01-01,100000,20세


### 4. 두 데이터 프레임의 모든 컬럼 이름이 다른 경우 

In [70]:
df_left = DataFrame({'이름':['영희', '철수'], '국어':[87, 91]})
df_left

,이름,국어
0,영희,87
1,철수,91


In [71]:
df_right = DataFrame({'성명':['영희', '철수'],'영어':[90,82]})
df_right

,성명,영어
0,영희,90
1,철수,82


In [72]:
# 왼쪽의 기준열 이름과 오른쪽의 기준을 이름을 각각 설정해야 한다. 
# 기준으로 지정된 모든 변수가 포함되므로, 
# 불필요한 변수는 drop()함수를 통해 삭제할 필요가 있다. 
r3 = merge(df_left, df_right, left_on=['이름'], right_on =['성명'])
r3

,이름,국어,성명,영어
0,영희,87,영희,90
1,철수,91,철수,82


## #04. 인덱스를 활용한 병합 

### 1. 학생의 이름을 인덱스로 갖는 두 데이터 프레임 생성 

In [73]:
df_left = DataFrame({'수학':[90, 82]}, index=['민철', '봉구'])
df_left

,수학
민철,90
봉구,82


In [74]:
df_right = DataFrame({'국어':[90,82]}, index=['민철', '철수'])
df_right

,국어
민철,90
철수,82


### 2. 병합 조건에서 index를 사용하도록 지정하여 Inner join하기 

In [75]:
merge(df_left, df_right, left_index = True, right_index=True)

,수학,국어
민철,90,90


### 3. 병합 조건에서 index를 사용하도록 지정하여 Left Outer Join 하기 

In [76]:
merge(df_left, df_right, left_index=True, right_index=True, how="left")


,수학,국어
민철,90,90.000
봉구,82,NaN


### 4. 병합 조건에서 index를 사용하도록 지정하여 Right Outer Join 하기 

In [77]:
merge(df_left, df_right, left_index=True, right_index=True, how="right")


,수학,국어
민철,90.000,90
철수,NaN,82


### 5. 병합 조건에서 index를 사용하도록 지정하여 Full Outer Join 하기

In [78]:
merge(df_left, df_right, left_index=True, right_index=True, how="outer")

,수학,국어
민철,90.000,90.000
봉구,82.000,NaN
철수,NaN,82.000


## #05. 인덱스와 컬럼을 각각 기준으로 하기 

### 1. 샘플 데이터 프레임 생성 

In [79]:
df_left = DataFrame({'수학':[90, 82]}, index=['민철', '봉구'])
df_left

,수학
민철,90
봉구,82


In [80]:
df_right = DataFrame({'성명':['민철', '철수'], '영어':[90, 82]})
df_right

,성명,영어
0,민철,90
1,철수,82


### 2. 왼쪽은 인덱스, 오른쪽은 컬럼을 기준으로 Inner Join 

In [81]:
merge(df_left, df_right, left_index=True, right_on=['성명'])


,수학,성명,영어
0,90,민철,90


### 3. 왼쪽은 인덱스, 오른쪽은 컬럼을 기준으로 Left Outer Join 

In [82]:
merge(df_left, df_right, left_index=True, right_on=['성명'], how='left')

,수학,성명,영어
0.000,90,민철,90.000
NaN,82,봉구,NaN


### 4. 왼쪽은 인덱스, 오른쪽은 컬럼을 기준으로 Right Outer Join 

In [83]:
merge(df_left, df_right, left_index=True, right_on=['성명'], how='right')

,수학,성명,영어
0,90.000,민철,90
1,NaN,철수,82


### 5. 왼쪽은 인덱스, 오른쪽은 컬럼을 기준으로 Full Outer Join

In [84]:
tmp = merge(df_left, df_right, left_index=True, right_on=['성명'], how='outer')
tmp

,수학,성명,영어
0.000,90.000,민철,90.000
NaN,82.000,봉구,NaN
1.000,NaN,철수,82.000


### 6. 병합 후에는 인덱스를 정리해야 한다.

In [85]:
tmp2 = tmp.set_index('성명')
tmp2

,수학,영어
성명,,
민철,90.000,90.000
봉구,82.000,NaN
철수,NaN,82.000


##  #06. 행 기준 병합 

### 1. 샘플 데이터 생성 

In [86]:
df1 = DataFrame({'이름':['영희', '철수'], '국어':[87, 91]})
df1

,이름,국어
0,영희,87
1,철수,91


In [87]:
df2 = DataFrame({'이름':['민철', '수현'],'국어':[78, 92]})
df2

,이름,국어
0,민철,78
1,수현,92


### 2. concat() 함수 기본 사용 방법

In [88]:
concat([df1, df2])

,이름,국어
0,영희,87
1,철수,91
0,민철,78
1,수현,92


### 3. 병합 결과에서 index를 재구성하기 (1)
- 지금까지 공부한 내용만으로 처리하기

In [89]:
인덱스_재구성df = concat([df1, df2])
인덱스_재구성df.reset_index(inplace=True, drop=True)
인덱스_재구성df

,이름,국어
0,영희,87
1,철수,91
2,민철,78
3,수현,92


### 4. 병합 결과에서 index를 재구성하기 (2)
- concat()함수에 ignore_index = True 파라미터를 설정하면 병합 후 인덱스를 재구성 한다.

In [90]:
concat([df1, df2], ignore_index=True)

,이름,국어
0,영희,87
1,철수,91
2,민철,78
3,수현,92


## #07. 서로 다른 변수를 갖는 데이터 프레임의 병합 

### 1. 샘플 데이터 구성 

In [91]:
df3 = DataFrame({'이름':['영희', '철수'], '국어':[87, 91]})
df3

,이름,국어
0,영희,87
1,철수,91


In [92]:
df4 = DataFrame({'이름':['민철', '수현'], '수학':[78, 92]}) 
df4

,이름,수학
0,민철,78
1,수현,92


### 2. 데이터 병합 
- 일치하지 않는 부분에 대해서는 NaN으로 처리한다. 

In [93]:
concat([df3, df4], ignore_index=True)

,이름,국어,수학
0,영희,87.000,NaN
1,철수,91.000,NaN
2,민철,NaN,78.000
3,수현,NaN,92.000


## #08. 인덱스를 갖는 데이터 프레임의 병합 

### 1. 샘플 데이터 구성 

In [94]:
df5 = DataFrame({'국어':[87, 91, 85]}, 
                index= ['영희', '철수', '민수'])
df5

,국어
영희,87
철수,91
민수,85


In [95]:
df6 = DataFrame({'국어':[78, 92, 82] }, 
                index=['민철', '수현', '민수'])
df6

,국어
민철,78
수현,92
민수,82


### 2. 데이터 병합 
- 인덱스도 함께 병합된다. 
- 각 데이터 프레임간에 중복되는 인덱스가 존재할 경우 각 행은 개별적으로 존재한다. (인덱스 중복)

In [96]:
concat([df5, df6])

,국어
영희,87
철수,91
민수,85
민철,78
수현,92
민수,82


### 3. 인덱스를 제외하고 병합하기 
- ignore_index = True 파라미터를 설정한다.

In [97]:
concat([df5, df6], ignore_index=True)

,국어
0,87
1,91
2,85
3,78
4,92
5,82


## #01. 연습문제 

In [98]:
import numpy as np
from jussam import load_data
from sklearn.impute import SimpleImputer

### 고객데이터(직접생성)

In [ ]:
# DataFrame 만드는 방법 3가지
# 방법 1: 딕셔너리 직접 넣기 (컬럼 기준)
# pythondf = DataFrame({
#     'class': ['A1', 'A1', 'A2'],
#     'time': [62, 60, 67]
# })

# 컬럼명이 키, 값들의 리스트가 값
# 한 번에 작성할 때 편함


# 방법 2: 딕셔너리를 변수로 분리
# pythondata = {
#     'class': ['A1', 'A1', 'A2'],
#     'time': [62, 60, 67]
# }
# df = DataFrame(data)

# 방법 1이랑 완전히 동일, 그냥 data 변수에 먼저 담은 것
# 데이터가 길면 이게 가독성 좋음


# 방법 3: 리스트 of 딕셔너리 (행 기준)
# pythondata = [
#     {'class': 'A1', 'time': 62},
#     {'class': 'A1', 'time': 60},
#     {'class': 'A2', 'time': 67}
# ]
# df = DataFrame(data)

# 행 하나가 딕셔너리 하나
# 방법 1, 2랑 결과는 동일한데 입력 방향이 반대


# 핵심 차이 한 줄 요약
# 구조방향방법 1, 2{컬럼: [값, 값, 값]}컬럼 기준방법 3[{컬럼: 값}, {컬럼: 값}]행 기준

In [ ]:
# 방법 1 (짧게)
pythoncustomers_df = DataFrame({
    'customer_id': [101, 102, 103, 104, 105, 106],
    'name': ['Alice', 'Bob', 'Charlie', 'David', 'Eve', 'Frank'],
    'age': [25, 32, 45, 38, 29, 41]
})


In [ ]:
# 방법 2 (분리)
pythoncustomer_data = {
    'customer_id': [101, 102, 103, 104, 105, 106],
    'name': ['Alice', 'Bob', 'Charlie', 'David', 'Eve', 'Frank'],
    'age': [25, 32, 45, 38, 29, 41]
}
customers_df = DataFrame(customer_data)

In [99]:
customer_data = {
    'customer_id': [101, 102, 103, 104, 105, 106],
    'name': ['Alice', 'Bob', 'Charlie', 'David', 'Eve', 'Frank'],
    'age': [25, 32, 45, 38, 29, 41]
}
customers_df = DataFrame(customer_data)
customers_df

,customer_id,name,age
0,101,Alice,25
1,102,Bob,32
2,103,Charlie,45
3,104,David,38
4,105,Eve,29
5,106,Frank,41


In [100]:
# DataFrame( 열고, 딕셔너리 { 열었는데, 딕셔너리 닫는 }를 빠뜨리고 )로 끝낸 거야.
# 외우는 팁: DataFrame({})  이 뼈대를 먼저 쳐놓고 안을 채워. 열고 바로 닫아두면 빠뜨릴 일이 없어.
# 고객정보 = DataFrame({'customer_id' : [101, 102, 103, 104, 105, 106], 'name': ['Alice', 'Bob', 'Charlie', 'David','Frank'], 'age' : [5, 32, 45, 38, 29, 41])

In [101]:
고객정보 = DataFrame({
    'customer_id': [101, 102, 103, 104, 105, 106],
    'name': ['Alice', 'Bob', 'Charlie', 'David', 'Eve','Frank'],
    'age': [25, 32, 45, 38, 29, 41]
})

### 구매 데이터 (데이터 가져오기)

In [103]:
purchases_df = load_data("order_history")
purchases_df

📚 신입 데이터 분석가를 위한 실전 미션!

    field        description
--  -----------  ----------------
 0  purchase_id  일련번호(인덱스)
 1  customer_id  고객번호
 2  item         상품명
 3  price        금액



,customer_id,item,price
purchase_id,,,
1,101,Laptop,1200.000
2,102,Mouse,25.000
3,103,Keyboard,NaN
4,101,Monitor,300.000
5,104,Webcam,5000.000
6,102,USB Hub,30.000
7,105,Laptop,1100.000
8,103,Mouse,NaN
9,101,External HDD,80.000


### 1. 구매 데이터의 price 컬럼에 있는 결측치는 전체 상품의 평균 가격으로 채운다.

In [104]:
결측치처리 = purchases_df.copy()
결측치처리

,customer_id,item,price
purchase_id,,,
1,101,Laptop,1200.000
2,102,Mouse,25.000
3,103,Keyboard,NaN
4,101,Monitor,300.000
5,104,Webcam,5000.000
6,102,USB Hub,30.000
7,105,Laptop,1100.000
8,103,Mouse,NaN
9,101,External HDD,80.000


In [102]:
### price컬럼에 있는 결측지 전체 상품의 평균 가격으로 채우기

In [105]:
average_price = 결측치처리['price'].mean()
average_price

np.float64(985.625)

In [106]:
결측치처리_결과 = 결측치처리['price'].fillna(average_price)
결측치처리_결과

purchase_id
1    1200.000
2      25.000
3     985.625
4     300.000
5    5000.000
6      30.000
7    1100.000
8     985.625
9      80.000
10    150.000
Name: price, dtype: float64

In [107]:
결측치처리['price'] = 결측치처리_결과
결측치처리

,customer_id,item,price
purchase_id,,,
1,101,Laptop,1200.000
2,102,Mouse,25.000
3,103,Keyboard,985.625
4,101,Monitor,300.000
5,104,Webcam,5000.000
6,102,USB Hub,30.000
7,105,Laptop,1100.000
8,103,Mouse,985.625
9,101,External HDD,80.000


### 2. 정제된 두 데이터 프레임을 customer_id 기준으로 병합하세요

In [116]:
병합 = merge(결측치처리, customers_df, on='customer_id')
병합

,customer_id,item,price,name,age
0,101,Laptop,1200.000,Alice,25
1,102,Mouse,25.000,Bob,32
2,103,Keyboard,985.625,Charlie,45
3,101,Monitor,300.000,Alice,25
4,104,Webcam,5000.000,David,38
5,102,USB Hub,30.000,Bob,32
6,105,Laptop,1100.000,Eve,29
7,103,Mouse,985.625,Charlie,45
8,101,External HDD,80.000,Alice,25
9,106,Docking Station,150.000,Frank,41


### 기술 통계량 도출 
- customer_id를 제외하고 기술통계량을 도출해야 한다.

In [117]:
기술통계_df = 병합.drop(columns=['customer_id', 'item', 'name']) 
기술통계_df.describe().T

,count,mean,std,min,25%,50%,75%,max
price,10.000,985.625,1490.666,25.000,97.500,642.812,1071.406,5000.000
age,10.000,33.700,8.042,25.000,26.000,32.000,40.250,45.000


## #02. 연습문제 - 흩어진 판매 데이터 통합하기 

### 데이터 가져오기

In [119]:
강남 = load_data('shop_강남')
강남

📚 흩어진 판매 데이터 통합하기(인덱스/메타데이터 없음)


,OrderID,Product,Quantity,Price
0,A001,Keyboard,10,15000.000
1,A002,Mouse,20,NaN
2,A003,Monitor,5,75000.000
3,A004,Keyboard,12,15000.000


In [109]:
종로 = load_data('shop_종로')
종로

📚 흩어진 판매 데이터 통합하기(인덱스/메타데이터 없음)


,OrderID,Product,Quantity,Price
0,B001,Mouse,15,8000
1,B002,Keyboard,999,14500
2,B003,Webcam,10,30000
3,B004,Mouse,18,8000


### 1. 강남점 데이터에는 'Mouse' 상품의 가격(Price)이 누락되어 있습니다. 이 값을 8000으로 채워넣기

In [ ]:
강남채우기 = 강남.copy()
강남채우기.loc[1, 'Price'] = 8000
강남채우기

,OrderID,Product,Quantity,Price
0,A001,Keyboard,10,15000.000
1,A002,Mouse,20,8000.000
2,A003,Monitor,5,75000.000
3,A004,Keyboard,12,15000.000


### 2. 종로점 데이터에는 Keyboard 상품의 주문 수량(Quantity)이 999로 잘못 입력되어 있습니다.
### 실제 수량은 9이므로, 이 값을 수정해주세요. 

In [111]:
종로채우기 = 종로.copy()
종로채우기.loc[1, 'Quantity'] = 9
종로채우기

,OrderID,Product,Quantity,Price
0,B001,Mouse,15,8000
1,B002,Keyboard,9,14500
2,B003,Webcam,10,30000
3,B004,Mouse,18,8000


In [112]:
종로 = 종로채우기.copy()
종로

,OrderID,Product,Quantity,Price
0,B001,Mouse,15,8000
1,B002,Keyboard,9,14500
2,B003,Webcam,10,30000
3,B004,Mouse,18,8000


In [113]:
강남 = 강남채우기.copy()
강남

,OrderID,Product,Quantity,Price
0,A001,Keyboard,10,15000.000
1,A002,Mouse,20,8000.000
2,A003,Monitor,5,75000.000
3,A004,Keyboard,12,15000.000


### 3, 4. 두 개의 DataFrame을 하나의 DataFrame으로 합치세요. 
### 합치는 과정에서 기존의 인덱스는 의미가 없으므로, 새로운 인덱스를 부여하세요. 

In [114]:
합치기 = concat([강남, 종로])
합치기

,OrderID,Product,Quantity,Price
0,A001,Keyboard,10,15000.000
1,A002,Mouse,20,8000.000
2,A003,Monitor,5,75000.000
3,A004,Keyboard,12,15000.000
0,B001,Mouse,15,8000.000
1,B002,Keyboard,9,14500.000
2,B003,Webcam,10,30000.000
3,B004,Mouse,18,8000.000


In [115]:
최종 = DataFrame(합치기)
최종

,OrderID,Product,Quantity,Price
0,A001,Keyboard,10,15000.000
1,A002,Mouse,20,8000.000
2,A003,Monitor,5,75000.000
3,A004,Keyboard,12,15000.000
0,B001,Mouse,15,8000.000
1,B002,Keyboard,9,14500.000
2,B003,Webcam,10,30000.000
3,B004,Mouse,18,8000.000
